### PR TITLE
BUG: sparse: Fix argmin/max shape diff between axis 0/1. And min/max can now return 1D shapes when axis set.

### PR DESCRIPTION
**Pull request by dschult**
Thus, this PR should go in before the 1.14 release or be backported if that doesn't happen.

This PR stems from an overview of the shapes returned by reduction methods (other than indexing).
Specifically, I've looked at `sum`, `mean`, `argmin`, `argmax`, `min`, `max`, `nanmin`, `nanmax` because they have an `axis` parameter. Let me know if there are others I should include (in another PR). Any reduction operation should be explored in light of the availability of 1D sparse to match the array-api.

For the default `axis=None`, all such functions return scalars for 1D, 2D sparray and 2D spmatrix. So that's good. But `argmin` and `argmax` return Python `int` rather than a numpy scalar. I have left that in place since the argmin should be an index. 

The methods `sum`, `mean`, `argmin`, and `argmax` return dense numpy objects when a valid `axis` is provided. For `spmatrix` the return is a 2D row or column depending on the axis value 0/1. For `sparray` the return is a 1D array. (The current release has `argmax(axis=0)` return 2D while `argmax(axis=1)` returns 1D. This PR fixes that for argmax and argmin.)

The other methods `min`, `max`, `nanmin`, `nanmax` return sparse matrix/array when `axis` is provided. For `spmatrix` they return a 2D row or column depending on the axis value 0/1. For `sparray` the return is a 1D object. (The current main returns a 2D row/column which breaks the array api. So this PR changes it to return 1D.)

Change summary:
- argmin/argmax now return 1D dense for axis=0/1/-1/-2 when input is a sparray.
- min/max/nanmin/nanmax now return 1D sparse for axis=0/1/-1/-2 when input is a sparray.
- tests are added to ensure this behavior is tracked and to adjust existing tests where needed. (I put these tests in `test_minmax1d.py` even though the input is 2D because the output is 1D.)

We've been wanting an overview of the current status of these functions, so I include a table here for discussion in case there are other aspects we wish to change (e.g. whether all these methods should return dense arrays).

This table shows the output of each method with given axis value as it depends on the input type/shape.
Note: All these methods return 0D numpy scalars when `axis=None` except `argmin/argmax` which return a Python `int` scalar.

Axis | Method | 1D-sparray   | 2D-sparray      | 2D-spmatrix
---  | --- | --- | --- | ---
0  |  sum   |  0D scalar  | 1D **np.ndarray** | 2D **np.matrix** row
0  |  mean |   0D scalar  | 1D **np.ndarray** | 2D **np.matrix** row
0  |  argmin | 0D python int | 1D **np.ndarray** | 2D **np.matrix** row
0  |  argmax | 0D python int | 1D **np.ndarray** | 2D **np.matrix** row
0  |  min    | 0D scalar   | 1D **sparray**  | 2D **spmatrix** row
0  |  max   |  0D scalar   | 1D **sparray** | 2D **spmatrix** row
0  |  nanmin | 0D scalar   | 1D **sparray** | 2D **spmatrix** row 
0  |  nanmax | 0D scalar   | 1D **sparray** | 2D **spmatrix** row 
1  |  sum     |  --       | 1D **np.ndarray** | 2D **np.matrix** col
1  |  mean   |   --       | 1D **np.ndarray** | 2D **np.matrix** col
1   | argmin   | --       | 1D **np.ndarray** | 2D **np.matrix** col
1  |  argmax  |  --       | 1D **np.ndarray** | 2D **np.matrix** col
1   | min     |  --       | 1D **sparray** | 2D **spmatrix** col
1   | max    |   --       | 1D **sparray** | 2D **spmatrix** col
1   | nanmin |   --       | 1D **sparray** | 2D **spmatrix** col
1   | nanmax  |  --       | 1D **sparray** | 2D **spmatrix** col

### CODE REVIEW

**Review by tylerjereddy — COMMENTED** [review]
> this PR should go in before the 1.14 release

Noted/milestoned. I should definitely let CJ review it properly.

It does look reasonably solid from a mechanical standpoint--old tests seem to get more thorough and a new test is added. I also checked all lines in the diff are covered by tests when I ran locally.

The one new test added here does seem to have quite a lot of things going on.

**Comment by tylerjereddy** [comment]
The `mypy` failure does look related

**Comment by tylerjereddy** [comment]
@perimosocordiae let me know if you think this will be reviewed/ready pretty soon; I could consider backporting during the RC stage if it is really important for `1.14.x` but needs more time.

File: scipy/_lib/cobyqa [hunk 1]
```diff
@@ -1 +1 @@
-Subproject commit c516987f947ccef9a16869b9633c20f9b8e0f4fe
+Subproject commit dee1a92b5f3fa50c6fce42a38a52ad2f4ebddc4c
```
Comment (perimosocordiae) [inline scipy/_lib/cobyqa:1]: Looks like this snuck in unintentionally

File: scipy/sparse/tests/test_minmax1d.py [hunk 2]
```diff
@@ -80,3 +80,58 @@ def test_argmax(self, spcreator):
...
+            (datspm.nanmax, np.nanmax),
+            (datspm.argmin, np.argmin),
+            (datspm.argmax, np.argmax),
+        ]:
+            assert_equal(toarray(spminax(axis=0)), npminax(dat, axis=0, keepdims=True))
+            assert_equal(toarray(spminax(axis=1)), npminax(dat, axis=1, keepdims=True))
+            assert_equal(spminax(axis=0).shape, (1, 4))
+            assert_equal(spminax(axis=1).shape, (3, 1))
+
+        assert datspm.min(axis=0).format == "coo"
```
Comment (perimosocordiae) [inline scipy/sparse/tests/test_minmax1d.py:126]: Add a loop for axis over (0, 1) here?

**Review by perimosocordiae — APPROVED** [review]
This is a backward compatibility break, but I'm okay to consider it a bug fix as it aligns us with the array API. Unfortunately there isn't really an opportunity to deprecate the old behavior here.

Behavior for spmatrix classes isn't changing.

**Review by tylerjereddy — APPROVED** [review]
CJ seems happy enough/comments addressed. CI is also happy, and my previous surface-level review comments were positive as well. There's a specific request to put this in `1.14.x` series, so let's try it.

**Pull request merged**